### PR TITLE
HSEARCH-3283 Add an extensive migration guide to the documentation to help users migrate from Search 5 to Search 6

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-3283

The goal was to have an extensive guide, and, well... it's extensive.

@fax4ever: maybe I should merge right away? Proof-reading this will take you a very long time...

### CODE REVIEW

**Comment by fax4ever** [comment]
I'm going to have a shallow look...

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 1]
```diff
@@ -0,0 +1,3351 @@
...
+        .within(51,Unit.KM)
+        .ofCoordinates(center)
+        .createQuery()
+|GeoPoint center = ...;
+f.spatial().within()
+        .field("field")
+        .circle(center,
+                51,
+                DistanceUnit.KILOMETERS)
+
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:2134]: extra-space

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 2]
```diff
@@ -0,0 +1,3351 @@
...
+Alternatively, `f.documentReference()` (without the wrapping in the "composite" projection)
+will return an instance of `DocumentReference`,
+which includes both the type and ID of the document.
+
+|EXPLANATION
+|// Lucene
+f.extension(LuceneExtension.get())
+        .explanation()
+// Elasticsearch
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:2417]: this line is not visible from my rendered page

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 3]
```diff
@@ -0,0 +1,3351 @@
...
+                .minDocumentCount( 1 )
+                .maxTermCount( 3 ) )
+        .fetch( 20 );
+
+List<Book> hits = result.hits();
+
+Map<Genre, Long> countByGenre = result.aggregation( countByGenreKey );
+----
+
+See [this section of the documentation] for more information.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:2527]: we don't have the link here

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 4]
```diff
@@ -0,0 +1,3351 @@
...
+`orderedBy`, `includeZeroCounts` and `maxFacetCount` have no equivalent Hibernate Search 6 range aggregations:
+all given ranges will always be included in the resulting `Map`.
+
+The behavior of these methods can be implemented by post-processing the `Map` in user code.
+====
+
+[[searching-facets-selectfacets]]
+===== Drill-down with `.selectFacets`
+
+In Hibernate Search 5, the `.selectFacets` method used to allow "drill-down",
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:2598]: s/used/was used/ ?

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 5]
```diff
@@ -0,0 +1,3351 @@
...
+
+See link:{hibernateSearchDocUrl}#mapper-orm-indexing-jsr352[this section of the documentation]
+for more information about the JSR-352 integration.
+
+See the table below for the exact equivalent of parameters that changed.
+
+.Job Parameters in JSR 352 Integration that changed in Hibernate Search 6
+|===
+|Hibernate Search 5 |Hibernate Search 6 .2+|Comment
+|Parameter Name |Builder Method |Parameter Name |Builder Method
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:2996]: repeated twice

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 6]
```diff
@@ -0,0 +1,3351 @@
...
+public class MySubClassEntity extends MyEntity {
+	// ...
+}
+----
+
+In Hibernate Search 6, `@Indexed` *is* inherited by subclasses;
+the configuration above would result in instances of both `MyEntity` and `MySubClassEntity` being indexed,
+though in different indexes.
+A search query on type `MyEntity` will return instances of both `MyEntity` and `MySubClassEntity`,
+in accordance with the https://en.wikipedia.org/wiki/Liskov_substitution_principle[Liskov substitution principle].
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:3096]: good!

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 7]
```diff
@@ -0,0 +1,3351 @@
...
+If it can't resolve it (for example when an `@IndexedEmbedded` association has no inverse side),
+Hibernate Search 6 will raise errors at bootstrap, looking like this:
+
+> Cannot find the inverse side of the association on type 'A' at path '.b<no value extractors>'
+> Hibernate Search needs this information in order to reindex 'A' when 'B' is modified.
+> You can solve this error by defining the inverse side of this association,
+> either with annotations specific to your integration (`@OneToMany(mappedBy = ...)` in Hibernate ORM)
+> or with the Hibernate Search `@AssociationInverseSide` annotation.
+> Alternatively, if you do not need to reindex 'A' when 'B' is modified,
+> you can disable automatic reindexing with `@IndexingDependency(reindexOnUpdate = ReindexOnUpdate.SHALLOW)`.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:3208]: the rendered of the quotation is not very nice IMO

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 8]
```diff
@@ -0,0 +1,3351 @@
...
+link:{hibernateSearchDocUrl}#mapper-orm-reindexing-derivedfrom[this section of the documentation].
+
+Alternatively, if you do not care about automatic reindexing when the transient property changes,
+then you can just disable automatic reindexing for this property with `@IndexingDependency(reindexOnUpdate = ReindexOnUpdate.NO)`.
+See link:{hibernateSearchDocUrl}#mapper-orm-reindexing-reindexonupdate-no[this section of the documentation].
+
+[[implicit-container-extraction]]
+=== Container extraction for `@*Field` is implied
+
+In Hibernate Search 6, when a `@*Field` annotation is applied on a property of a container type,
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/migration/index.asciidoc:3261]: we should escape the `*` here

**Review by fax4ever — APPROVED** [review]
Very nice and extensive job.
I couldn't resist to review it.
Anyway I've few comments + I was thinking that eventually we should describe the fact that now custom bridges cannot define properties at indexing time (e.g.: ISPN integration).

File: documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmManualIndexingIT.java [hunk 9]
```diff
@@ -198,12 +200,22 @@ public void workspace() {
 		initBooksAndAuthors( entityManagerFactory, numberOfBooks );
 
 		OrmUtils.withinEntityManager( entityManagerFactory, entityManager -> {
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmManualIndexingIT.java:202]: Good catch: we do not. Fixed.

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 10]
```diff
@@ -0,0 +1,3351 @@
...
+== Migration helper
+
+=== Purpose
+
+Hibernate Search 6 includes a temporary additional "migration helper" module
+that provides partial compatibility with Hibernate Search 5 APIs backed by the Hibernate Search 6 implementations.
+
+This module should make migration easier by making sure that code relying on the most-frequently-used APIs
+(mapping annotations, search DSL, ...)
+continues to compile and run.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/migration/index.asciidoc:94]: The code continues, not the APIs.

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 11]
```diff
@@ -0,0 +1,3351 @@
...
++
+Hibernate Search 6 equivalent (per-index): none.
++
+Setting `hibernate.search.automatic_indexing.synchronization.strategy` to `read-sync` or `sync` will produce results
+similar to setting `hibernate.search.default.elasticsearch.refresh_after_write` to `true`.
+See link:{hibernateSearchDocUrl}#mapper-orm-indexing-automatic-synchronization[automatic indexing synchronization]
+for more information.
++
+There is no equivalent for the per-index variant in Hibernate Search 6:
+the synchronization strategy can only be set globally, not on a per-index basis.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/migration/index.asciidoc:284]: This is about an Elasticsearch setting; `refresh_interval` is only exposed for the Lucene backend. Granted, there are ways to configure something similar through Elasticsearch index settings, but I'd rather not go into these details here.

File: documentation/src/main/asciidoc/migration/index.asciidoc [hunk 12]
```diff
@@ -0,0 +1,3351 @@
...
+----
+@KeywordField(normalizer = "myNormalizer")
+----
+|[source, JAVA]
+----
+@Field(analyze = analyze.NO)
+----
+|[source, JAVA]
+----
+@KeywordField
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/migration/index.asciidoc:773]: Either would work. I'll mention it.

File: documentation/src/main/asciidoc/migration/index.asciidoc (hunk 2 above)
Comment (yrodiere) [reply documentation/src/main/asciidoc/migration/index.asciidoc:2417]: Fixed

File: documentation/src/main/asciidoc/migration/index.asciidoc (hunk 3 above)
Comment (yrodiere) [reply documentation/src/main/asciidoc/migration/index.asciidoc:2527]: Fixed

File: documentation/src/main/asciidoc/migration/index.asciidoc (hunk 4 above)
Comment (yrodiere) [reply documentation/src/main/asciidoc/migration/index.asciidoc:2598]: "used to" as in "in the past we allowed the drill-down operation, but now we don't allow it anymore".

**Comment by yrodiere** [comment]
Thanks! I think I addressed your comments. Merging...

**Pull request merged**